File: apps/desktop/src/app/notes/money.ts
/**
 * Money model: transactions live in the vault as markdown tables, one note
 * per month (`Money/2026-07.md`). Plain files the user owns and can edit by
 * hand; the agent appends rows after extracting them from a statement.
 *
 * Row format (GFM table):
 *   | Date | Description | Category | Amount |
 *   | 2026-07-14 | Adobe | Software | -89.99 |
 *
 * Amounts are signed decimals in the vault's single currency: negative is
 * money out. Parsing tolerates $, commas, and parenthesised negatives.
 */

export interface Transaction {
  date: string
  description: string
  category: string
  amount: number
  /** Vault path of the month note this row lives in. */
  path: string
  line: number
}

export const MONEY_DIR = 'Money'
const HEADER = '| Date | Description | Category | Amount |'
const DIVIDER = '| --- | --- | --- | --- |'

export function monthNotePath(date: string): string {
  return `${MONEY_DIR}/${date.slice(0, 7)}.md`
}

export function parseAmount(raw: string): number | null {
  const text = raw.trim().replace(/[$,\s]/g, '')

  if (!text) {
    return null
  }

  const negative = /^\(.*\)$/.test(text)
  const digits = text.replace(/[()]/g, '')

  if (!/^[-+]?\d*\.?\d+$/.test(digits)) {
    return null
  }

  const value = Number(digits)

  return negative ? -Math.abs(value) : value
}

export function formatAmount(value: number, currency = '$'): string {
  const sign = value < 0 ? '-' : '+'

  return `${sign}${currency}${Math.abs(value).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 })}`
}

/** Parse the transaction rows out of a month note. */
export function parseMonthNote(path: string, content: string): Transaction[] {
  const rows: Transaction[] = []
  const lines = content.split('\n')

  for (let index = 0; index < lines.length; index++) {
    const line = lines[index].trim()

    if (!line.startsWith('|') || line.startsWith('| ---') || /^\|\s*Date\s*\|/i.test(line)) {
      continue
    }

    const cells = line.split('|').slice(1, -1).map(cell => cell.trim())

    if (cells.length < 4) {
      continue
    }

    const [date, description, category, amountRaw] = cells
    const amount = parseAmount(amountRaw)

    if (!/^\d{4}-\d{2}-\d{2}$/.test(date) || amount === null) {
      continue
    }

    rows.push({ date, description, category: category || 'Uncategorized', amount, path, line: index + 1 })
  }

  return rows
}

/** Content for a month note that doesn't exist yet. */
export function emptyMonthNote(month: string): string {
  return `---\ntype: money\nmonth: ${month}\n---\n\n# ${month}\n\n${HEADER}\n${DIVIDER}\n`
}

/**
 * Append rows to a month note's table, creating the table if needed and
 * skipping rows already present (same date + amount + description).
 */
export function appendRows(content: string, rows: Transaction[]): { content: string; added: number; skipped: number } {
  const existing = new Set(
    parseMonthNote('', content).map(row => `${row.date}|${row.amount}|${row.description.toLowerCase()}`)
  )

  let next = content

  if (!next.includes(HEADER)) {
    next = next.trimEnd() + `\n\n${HEADER}\n${DIVIDER}\n`
  }

  let added = 0
  let skipped = 0
  const lines: string[] = []

  for (const row of rows) {
    const key = `${row.date}|${row.amount}|${row.description.toLowerCase()}`

    if (existing.has(key)) {
      skipped += 1
      continue
    }

    existing.add(key)
    added += 1
    lines.push(
      `| ${row.date} | ${row.description.replace(/\|/g, '/')} | ${row.category.replace(/\|/g, '/')} | ${row.amount.toFixed(2)} |`
    )
  }

  if (lines.length) {
    next = next.trimEnd() + '\n' + lines.join('\n') + '\n'
  }

  return { content: next, added, skipped }
}

export interface MoneySummary {
  income: number
  spend: number
  net: number
  byCategory: Array<{ category: string; total: number }>
}

export function summarize(rows: Transaction[]): MoneySummary {
  let income = 0
  let spend = 0
  const categories = new Map<string, number>()

  for (const row of rows) {
    if (row.amount >= 0) {
      income += row.amount
    } else {
      spend += Math.abs(row.amount)
    }

    categories.set(row.category, (categories.get(row.category) ?? 0) + row.amount)
  }

  return {
    income,
    spend,
    net: income - spend,
    byCategory: [...categories.entries()]
      .map(([category, total]) => ({ category, total }))
      .sort((a, b) => a.total - b.total)
  }
}

File: plugins/vault/money.py
"""Money tools: the agent's side of the statement-to-transactions flow.

The model reads a bank statement (image or PDF the user dropped into the
chat) and calls `money_add_transactions` with structured rows. We append
them to the month note as a markdown table — plain files the user owns —
skipping duplicates so re-importing the same statement is safe.
"""

from __future__ import annotations

import json
import re
from pathlib import Path

from .tools import _resolve, _vault_root, vault_write

HEADER = "| Date | Description | Category | Amount |"
DIVIDER = "| --- | --- | --- | --- |"
DATE_RE = re.compile(r"^\d{4}-\d{2}-\d{2}$")


def _month_path(date: str) -> str:
    return f"Money/{date[:7]}.md"


def _empty_note(month: str) -> str:
    return f"---\ntype: money\nmonth: {month}\n---\n\n# {month}\n\n{HEADER}\n{DIVIDER}\n"


def _existing_keys(content: str) -> set[str]:
    keys = set()

    for line in content.splitlines():
        stripped = line.strip()

        if not stripped.startswith("|") or stripped.startswith("| ---"):
            continue

        cells = [cell.strip() for cell in stripped.split("|")[1:-1]]

        if len(cells) < 4 or not DATE_RE.match(cells[0]):
            continue

        try:
            amount = float(cells[3].replace("$", "").replace(",", ""))
        except ValueError:
            continue

        keys.add(f"{cells[0]}|{amount:.2f}|{cells[1].lower()}")

    return keys


def money_add_transactions(rows_json: str, source: str = "") -> str:
    """Append extracted transactions to their month notes.

    rows_json: JSON array of {date: YYYY-MM-DD, description, category, amount}
    where amount is a signed number (negative = money out).
    """
    root = _vault_root()

    if not root:
        return "No vault is connected — ask the user to open a vault in BISEO first."

    try:
        rows = json.loads(rows_json) if isinstance(rows_json, str) else rows_json
    except (TypeError, ValueError):
        return "rows must be a JSON array of {date, description, category, amount}."

    if not isinstance(rows, list) or not rows:
        return "No transactions were provided."

    by_month: dict[str, list[dict]] = {}
    rejected = 0

    for row in rows:
        if not isinstance(row, dict):
            rejected += 1
            continue

        date = str(row.get("date", "")).strip()
        description = str(row.get("description", "")).strip()

        try:
            amount = float(str(row.get("amount", "")).replace("$", "").replace(",", ""))
        except (TypeError, ValueError):
            rejected += 1
            continue

        if not DATE_RE.match(date) or not description:
            rejected += 1
            continue

        by_month.setdefault(_month_path(date), []).append(
            {
                "date": date,
                "description": description.replace("|", "/"),
                "category": (str(row.get("category", "")).strip() or "Uncategorized").replace("|", "/"),
                "amount": amount,
            }
        )

    if not by_month:
        return f"None of the {len(rows)} rows were usable (need date YYYY-MM-DD, description, numeric amount)."

    added_total = 0
    skipped_total = 0
    touched = []

    for rel_path, month_rows in by_month.items():
        absolute = _resolve(root, rel_path)

        if not absolute:
            continue

        month = Path(rel_path).stem
        content = absolute.read_text(encoding="utf-8") if absolute.exists() else _empty_note(month)

        if HEADER not in content:
            content = content.rstrip() + f"\n\n{HEADER}\n{DIVIDER}\n"

        keys = _existing_keys(content)
        lines = []

        for row in sorted(month_rows, key=lambda item: item["date"]):
            key = f'{row["date"]}|{row["amount"]:.2f}|{row["description"].lower()}'

            if key in keys:
                skipped_total += 1
                continue

            keys.add(key)
            added_total += 1
            lines.append(
                f'| {row["date"]} | {row["description"]} | {row["category"]} | {row["amount"]:.2f} |'
            )

        if lines:
            content = content.rstrip() + "\n" + "\n".join(lines) + "\n"

            if source:
                marker = f"\n<!-- imported from: {source} -->\n"

                if marker.strip() not in content:
                    content = content.rstrip() + "\n" + marker

            vault_write(rel_path, content)
            touched.append(rel_path)

    summary = [f"Added {added_total} transaction(s)"]

    if skipped_total:
        summary.append(f"skipped {skipped_total} duplicate(s)")

    if rejected:
        summary.append(f"ignored {rejected} unusable row(s)")

    summary.append(f'in {", ".join(touched) if touched else "no files"}')

    return " · ".join(summary) + ". The user can review the table in the Money screen."


def money_summary(month: str = "") -> str:
    """Totals for a month (YYYY-MM); defaults to the latest month note."""
    root = _vault_root()

    if not root:
        return "No vault is connected."

    money_dir = root / "Money"

    if not money_dir.is_dir():
        return "No money notes yet. Drop a bank statement into the chat and I'll extract the transactions."

    if month.strip():
        target = money_dir / f"{month.strip()}.md"
    else:
        notes = sorted(money_dir.glob("*.md"))
        target = notes[-1] if notes else None

    if not target or not target.exists():
        return f"No note for {month or 'that month'}."

    content = target.read_text(encoding="utf-8")
    income = 0.0
    spend = 0.0
    categories: dict[str, float] = {}
    count = 0

    for line in content.splitlines():
        stripped = line.strip()

        if not stripped.startswith("|") or stripped.startswith("| ---") or stripped.lower().startswith("| date"):
            continue

        cells = [cell.strip() for cell in stripped.split("|")[1:-1]]

        if len(cells) < 4 or not DATE_RE.match(cells[0]):
            continue

        try:
            amount = float(cells[3].replace("$", "").replace(",", ""))
        except ValueError:
            continue

        count += 1

        if amount >= 0:
            income += amount
        else:
            spend += abs(amount)

        categories[cells[2] or "Uncategorized"] = categories.get(cells[2] or "Uncategorized", 0.0) + amount

    lines = [
        f"{target.stem}: {count} transaction(s)",
        f"  in  +{income:,.2f}",
        f"  out -{spend:,.2f}",
        f"  net {income - spend:+,.2f}",
        "  by category:",
    ]

    for category, total in sorted(categories.items(), key=lambda item: item[1]):
        lines.append(f"    {category}: {total:+,.2f}")

    return "\n".join(lines)
